content/articles/optimizing-ml-models-with-optuna/index.md

```diff
@@ -0,0 +1,155 @@
+```python
+import optuna
+# The Samplers class defines our hyper-parameter space.
+from Optuna Import Trial, visualization
+from optuna.samplers import TPESampler
+```
+#### Creating the primary objective function
+We place all the possible hyper-parameter ranges, and the method accepts a trial object and should return a score value.
+```python
+def objective(trial):
+  parameter = {
+                "n_estimators" : trial.suggest_int('n_estimators', 0, 500),
+                'subsample':trial.suggest_uniform('subsample',0.4,0.9),
+                'gamma':trial.suggest_uniform('gamma', 0, 4),
+                'reg_alpha':trial.suggest_uniform('reg_alpha',0,6),
+               #uniformly distributed numbers between 0,2
+                'reg_lambda':trial.suggest_uniform('reg_lambda',0,2),
+                'min_child_weight':trial.suggest_int('min_child_weight',0,5),
+                'learning_rate':trial.suggest_loguniform('learning_rate',0.05,0.5),
+               #log distribution between [0.05,.5]
+                'colsample_bytree':trial.suggest_uniform('colsample_bytree',0.4,0.9),
+                'max_depth':trial.suggest_int('max_depth', 3, 5),
+                'nthread' : -1
+            }
+  #returns the regressionModelse score
+  return(return_score(parameter))
+```
+#### Creating an optuna–study object
+This study object stores all the information about the hyper-parameters tested under that research.
+```python
+studyObject1 = optuna.create_study(
+    direction='minimize',sampler=TPESampler())
+    # Bayesian Sampling Technique
+studyObject1.optimize(objective, n_trials= 500)
+```
+After running i found that my best trial was 459.
+
+Output:
+
+![Best trial](/engineering-education/optimizing-ml-models-with-optuna/trial.png)
+#### Visualizing search history
+Our model can readily be shown to be affected by hyper-parameters by looking at the search history. It is possible to further decrease the scope of our search by cutting down the search parameters.
+```python
+optuna.visualization.plot_slice(studyObject1)
+```
+#### Comparing the optimized model with the base model
+However, we provide `studyObject1.best_parameter` to the optimized model, which generates a dictionary of the best hyper-parameters. The default model has no hyper-parameters.
+
+The object of study encompasses all of the information about a certain inquiry or study. Best parameters are returned by `studyObject1.best_params.`
+```python
+studyObject1.best_params
+```
+Output:
+```bash
+{'colsample_bytree': 0.8228985622676791,
+ 'gamma': 0.5064888131479657,
+ 'max_depth': 4,
+ 'min_child_weight': 1,
+ 'n_estimators': 475,
+ 'ourlearning_rate': 0.053232877076795374,
+ 'reg_alpha': 1.908137784762065,
+ 'reg_lambda': 0.03170430168340592,
+ 'subsample': 0.7131725068799382}
+ ```
+Let's examine how far we've come with our model:
+```python
+parameter = {}
+print(f"without optimization {return_score(parameter)}")

```

Comment:
Need more explanation. I can't really get the context of what is being explained.